File: station-agent/StationAgent.Service/Realtime/AgentProtocol.cs
using System.Text.Json;

namespace StationAgent.Service.Realtime;

internal static class AgentProtocol
{
    public const int Version = 1;

    public const string ConnectedType =
        "CONNECTED";

    public const string HeartbeatType =
        "HEARTBEAT";

    public const string HeartbeatAckType =
        "HEARTBEAT_ACK";

    public const string SessionStartType =
        "SESSION_START";

    public const string SessionExtendType =
        "SESSION_EXTEND";

    public const string SessionFinishType =
        "SESSION_FINISH";

    public const string ErrorType =
        "ERROR";

    public static JsonSerializerOptions
        JsonOptions { get; } = new()
        {
            PropertyNamingPolicy =
                JsonNamingPolicy.SnakeCaseLower,
        };
}


internal sealed record AgentHeartbeatMessage(
    int Version,
    string Type,
    Guid EventId,
    Guid? CorrelationId,
    DateTimeOffset SentAt
);


internal sealed record ServerAgentMessage(
    int Version,
    string Type,
    Guid EventId,
    Guid? CorrelationId,
    DateTimeOffset SentAt,
    JsonElement Data
);


internal sealed record ConnectedMessageData(
    Guid StationId,
    string StationCode,
    int HeartbeatIntervalSeconds,
    AgentSessionData? ActiveSession
);


internal sealed record AgentSessionData(
    Guid SessionId,
    string SessionType,
    int AuthorizedSeconds,
    DateTimeOffset StartedAt,
    DateTimeOffset ServerNow,
    int ElapsedSeconds,
    int RemainingSeconds,
    string TimeState
);


internal sealed record SessionFinishData(
    Guid SessionId,
    string SessionType,
    DateTimeOffset EndedAt
);


internal sealed record ServerErrorData(
    string? Code
);

File: app/schemas/agent_protocol.py
from datetime import datetime, timezone
from typing import Any, Literal
from uuid import UUID, uuid4

from pydantic import (
    BaseModel,
    ConfigDict,
    Field,
)


class AgentHeartbeatMessage(BaseModel):
    model_config = ConfigDict(
        extra="forbid",
    )

    version: Literal[1]
    type: Literal["HEARTBEAT"]
    event_id: UUID
    correlation_id: UUID | None = None
    sent_at: datetime


class ServerAgentMessage(BaseModel):
    model_config = ConfigDict(
        extra="forbid",
    )

    version: Literal[1] = 1

    type: Literal[
        "CONNECTED",
        "HEARTBEAT_ACK",
        "SESSION_START",
        "SESSION_EXTEND",
        "SESSION_FINISH",
        "ERROR",
    ]

    event_id: UUID = Field(
        default_factory=uuid4,
    )

    correlation_id: UUID | None = None

    sent_at: datetime = Field(
        default_factory=lambda: datetime.now(
            timezone.utc
        )
    )

    data: dict[str, Any] = Field(
        default_factory=dict,
    )